app/src/ui/history/merge-call-to-action-with-conflicts.tsx

```diff
@@ -0,0 +1,140 @@
+import * as React from 'react'
+
+import { CompareActionKind, MergeResultStatus } from '../../lib/app-state'
+import { Repository } from '../../models/repository'
+import { Branch } from '../../models/branch'
+import { Dispatcher } from '../../lib/dispatcher'
+import { Button } from '../lib/button'
+import { MergeStatusHeader } from './merge-status-header'
+import { MergeResultKind } from '../../models/merge'
+
+interface IMergeCallToActionWithConflictsProps {
+  readonly repository: Repository
+  readonly dispatcher: Dispatcher
+  readonly mergeStatus: MergeResultStatus | null
+  readonly currentBranch: Branch
+  readonly comparisonBranch: Branch
+  readonly commitsBehind: number
+
+  /**
+   * Callback to execute after a merge has been performed
+   */
+  readonly onMerged: () => void
+}
+
+export class MergeCallToActionWithConflicts extends React.Component<
+  IMergeCallToActionWithConflictsProps,
+  {}
+> {
+  public render() {
+    const { commitsBehind } = this.props
+
+    const cannotMergeBranch =
+      this.props.mergeStatus != null &&
+      this.props.mergeStatus.kind === MergeResultKind.Invalid
+
+    const disabled = commitsBehind <= 0 || cannotMergeBranch
+
+    return (
+      <div className="merge-cta">
+        <Button type="submit" disabled={disabled} onClick={this.onMergeClicked}>
+          Merge into <strong>{this.props.currentBranch.name}</strong>
+        </Button>
+
+        <div className="merge-status-component">
+          <MergeStatusHeader status={this.props.mergeStatus} />
+
+          {this.renderMergeDetails(
+            this.props.currentBranch,
+            this.props.comparisonBranch,
+            this.props.mergeStatus,
+            commitsBehind
+          )}
+        </div>
+      </div>
+    )
+  }
+
+  private renderMergeDetails(
+    currentBranch: Branch,
+    comparisonBranch: Branch,
+    mergeStatus: MergeResultStatus | null,
+    behindCount: number
+  ) {
+    if (mergeStatus === null || mergeStatus.kind === MergeResultKind.Loading) {
+      return (
+        <div className="merge-message merge-message-loading">
+          Checking for ability to merge automatically...
+        </div>
+      )
+    }
+    const branch = comparisonBranch
+
+    if (mergeStatus.kind === MergeResultKind.Clean) {
```

Comment:
i think this section might be easier to read (and maintain) if we used a switch case around `mergeStatus.kind` that returned a function that did the rendering of this section of the UI. similar to https://github.com/desktop/desktop/pull/5495/files#diff-127ff68ee7a894bbcaaa8c6e706a727bR40

Reply:
or maybe there's another way to structure this logic so its easier to read?